api/data/reviews.js

```diff
@@ -0,0 +1,206 @@
+const { ObjectId } = require('mongodb');
+const mongoCollections = require('../config/mongoCollections');
+const reviews = mongoCollections.reviews;
+const users = mongoCollections.users;
+const games = mongoCollections.games;
+
+const userMethods = require('../data/users');
+const gameMethods = require('../data/games');
+const validate = require('./validation');
+
+module.exports = {
+    // returns an array of all reviews
+    // if there are no reviews return an empty array
+    async getAllReviews() {
+        const reviewCollection = await reviews();
+        const reviewList = await reviewCollection.find.toArray();
+        return reviewList;
+    },
+
+    async getReviewById(id) {
+        if (arguments.length !== 1) throw "Usage: Review Id";
+        if (!ObjectId.isValid(id)) throw "Review Id needs to be a valid ObjectId";
+
+        const reviewCollection = await reviews();
+        const review = await reviewCollection.findOne({ _id: id });
+        if (!review) throw "Review not found with the given id";
+        return review;
+    },
+
+    //gets all reviews beloning to a user id
+    async getAllReviewsByUserId(id) {
+        if (arguments.length !== 1) throw "Usage: User Id";
+        if (!ObjectId.isValid(id)) throw "Review Id needs to be a valid ObjectId";
+
+        const reviewCollection = await reviews();
+        const reviews = await reviewCollection.find({userId: id}).toArray();
+        return reviews;
+    },
+
+    // it adds a review obviously
+    // comment is an optional parameter
+    async addReview(userId, gameId, rating, date, comment) {
+        if (arguments.length !== 4 && arguments.length !== 5)
+            throw "Usage: User Id, Game Id, Rating, Date, (Optional) Comment";
+        if (!Number.isInteger(rating) || rating < 1 || rating > 10)
+            throw "Rating needs to be a positive integer from 1-10";
+        if (comment && !validate.validateString(comment))
+            throw "If you supply a comment, its gotta be a non empty string";
+
+        let word = "";
+        if (comment)
+            word = comment;
+
+        // TODO: validate date
+
+        let user;
+        let game;
+
+        try {
+            user = await userMethods.getUserById(userId);
+            game = await gameMethods.getGameById(gameId);
+        } catch (e) {
+            throw e;
+        }
+
+        const newReview = {
+            _id: ObjectId(),
+            userId: userId,
+            gameId: gameId,
+            rating: rating,
+            comment: word,
+            displayName: user.displayName,
+            datePosted: date
+        };
+
+        const reviewCollection = await reviews();
+        const insertInfo = await reviewCollection.insertOne(newReview);
+        if (insertInfo.insertedCount === 0)
+            throw "Could not create new review object";
+
+        // need to update user and game review id lists
+        let userReviews = user.reviews;
```

Comment:
This section could be sorted a lot by using the built-in Mongo array push: https://docs.mongodb.com/manual/reference/operator/update/push/